theano/gpuarray/ROIPoolGPUFwd.c

```diff
@@ -0,0 +1,140 @@
+            ga_float bin_size_w = static_cast<ga_float>(roi_width) / static_cast<ga_float>(pooled_width);
+            for (ga_int c = 0; c < channels; ++c) {
+                ga_int data_inc = c * height * width;
+                ga_int out_channel_inc = c * pooled_height * pooled_width;
+                DTYPE_i0* channel_data = batch_data + data_inc;
+                DTYPE_o0* channel_out = roi_out + out_channel_inc;
+                DTYPE_o1* channel_argmax = roi_argmax + out_channel_inc;
+                for (ga_int ph = 0; ph < pooled_height; ++ph) {
+                    for (ga_int pw = 0; pw < pooled_width; ++pw) {
+                        ga_int hstart = static_cast<ga_int>(floor(static_cast<ga_float>(ph) * bin_size_h)) + roi_start_h;
+                        ga_int wstart = static_cast<ga_int>(floor(static_cast<ga_float>(pw) * bin_size_w)) + roi_start_w;
+                        ga_int hend = static_cast<ga_int>(ceil(bin_size_h)) + hstart;
+                        ga_int wend = static_cast<ga_int>(ceil(bin_size_w)) + wstart;
+                        // Add roi offsets and clip to input boundaries
+                        hstart = min(max(hstart, 0), height);
+                        hend = min(max(hend, 0), height);
+                        wstart = min(max(wstart, 0), width);
+                        wend = min(max(wend, 0), width);
+                        bool is_empty = (hend <= hstart) || (wend <= wstart);
+                        ga_int pool_index = ph * pooled_width + pw;
+                        if (is_empty) {
+                            channel_out[pool_index] = 0;
+                            channel_argmax[pool_index] = -1;
+                        }
+                        // Define an empty pooling region to be zero
+                        // If nothing is pooled, argmax = -1 causes nothing to be backprop'd
+                        for (ga_int h = hstart; h < hend; ++h) {
+                            for (ga_int w = wstart; w < wend; ++w) {
+                                ga_int bottom_index = h * width + w;
+                                if (channel_data[bottom_index] > channel_out[pool_index]) {
+                                    channel_out[pool_index] = channel_data[bottom_index];
+                                    channel_argmax[pool_index] = bottom_index;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#section support_code_struct
+
+bool vector_same_shape(PyGpuArrayObject* arr1, PyGpuArrayObject* arr2){
+  return (PyGpuArray_DIMS(arr1)[0] == PyGpuArray_DIMS(arr2)[0]);
+  }
+
+int APPLY_SPECIFIC(ROIPoolGPUFwd)(PyGpuArrayObject *data,
+                      PyGpuArrayObject *rois,
+                      PyGpuArrayObject **out,
+		              PyGpuArrayObject **argmaxes,
+                      PyGpuContextObject *ctx) {
+    size_t address = 1;
```

Comment:
The realization came when I was writing the pythonic implementation. The way i used to slice the image array to get the region from where i have to apply a max pool was, 
` image_array[b_in, c, y1:y2, x1:x2]`
(where `b_in` is the batch id)
Later my confusion arose because I had thought it should be done like(across the region in all the images in a batch, which is why i thought it was absurd :stuck_out_tongue:  )
`image_array[:, c, y1:y2, x1:x2]`
The correct way of doing it is, (which I realized it early morning, when I made those series of comments)
```
interested_batchid = [j for j in [rois[i][0] for i in range(len(rois))]]
image_array[interested_batchid[k], cn , x1:x2, y1:y2]
```
Am i correct now ?